File: aoc23swift/aoc23swift/puzzle.swift
//
// Created by Lee Christie on 2023-12-07.
//

import Foundation

enum InputError: Error {
    case UnableToParseLine
}

func readInput(day: Int, example: Bool = false) -> String {
    if (day < 1 || day > 25) {
        print("Unable to open input file.")
        exit(1)
    }
    let dayFormat = String(format: "%02d", day)
    let filename = "/Users/0x1ac/Developer/leechristie/advent-of-code-2023/input/"
    let prefix = example ? "example" : "input"
    let fullPath = "\(filename)\(prefix)\(dayFormat).txt"
    let rv = try? String(contentsOfFile: fullPath)
    guard let rv else {
        print("Unable to open input file.")
        exit(1)
    }
    return rv
}

File: aoc23go/puzzle.go
// Advent of Code 2023
// Dr Lee A. Christie
// @0x1ac@techhub.social

package main

import (
	"io"
	"os"
	"strconv"
	"strings"
)

func OpenInputFile(day int) (*os.File, func(), error) {
	var dayStr = strconv.Itoa(day)
	if day < 10 {
		dayStr = "0" + dayStr
	}
	var filename = "/Users/0x1ac/Developer/leechristie/advent-of-code-2023/input/input" + dayStr + ".txt"
	file, err := os.Open(filename)
	if err != nil {
		return nil, nil, err
	}
	var closer = func() {
		file.Close()
	}
	return file, closer, nil
}

func ReadInputFile(day int) (string, error) {
	file, closer, err := OpenInputFile(1)
	if err != nil {
		return "", err
	}
	defer closer()
	reader := io.Reader(file)
	data, err := io.ReadAll(reader)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func ReadInputLines(day int) ([]string, error) {
	lines, err := ReadInputFile(1)
	if err != nil {
		return nil, err
	}
	return strings.Split(lines, "\n"), nil
}
